### PR TITLE
perf(retrieval): optimize BM25 scoring hot path

### PR DESCRIPTION
**Pull request by d-o-hub**
This PR implements two key algorithmic improvements to the BM25 retrieval module:
1. **Pre-computed weighted IDF**: `idf * (k1 + 1)` is now computed once per query instead of per-document.
2. **Hoisted den_base constant**: Document-length normalization denominator base is calculated once per document using `f32::mul_add` for hardware acceleration.

Additionally:
- `add_document` is now generic over `AsRef<str>`, supporting both string slices and owned collections.
- Added unit tests for zero-length documents and query edge cases.
- Verified ~4% performance gain in search benchmarks.

---
*PR created automatically by Jules for task [9294260772045228186](https://jules.google.com/task/9294260772045228186) started by @d-o-hub*

### CODE REVIEW

**Comment by d-o-hub** [comment]
@jules 
Objective:Ensure that all Continuous Integration (CI) issues are identified, addressed, and resolved, and that all open conversations in pull requests or code reviews are properly handled and closed.
Instructions:

Review CI Status and Logs:  

Check the CI pipeline status for all recent commits and pull requests.  
Identify any failing jobs, tests, or quality checks reported by the CI system (e.g., build failures, test failures, linting errors, security scans).  
Review detailed logs to understand the root cause of each failure.


Fix CI Issues:  

For each identified CI failure, apply the necessary code fixes, configuration updates, or dependency adjustments.  
Ensure that all automated tests pass successfully.  
Verify that code quality and security checks meet the required standards.


Run CI Locally (if applicable):  

Use local CI runners or tools (e.g., act for GitHub Actions) to reproduce and debug CI issues before pushing fixes.  
Confirm that the fixes resolve the issues locally.


Address Open Conversations:  

Review all open conversations in the pull request or code review threads.  
Respond to comments with clarifications, explanations, or code updates as needed.  
Make any requested changes or improvements to the code.  
Mark conversations as resolved once the concerns have been addressed.


Final Verification:  

After applying fixes and resolving conversations, push the changes and trigger a new CI run.  
Confirm that the CI pipeline completes successfully without errors.  
Ensure that all conversations are closed and the pull request is ready for merge.

**Comment by d-o-hub** [comment]
@jules compact / update learnings and check plans/ - agents.md and reference files if changes are needed. use dog food skill and verify in db and files save/load - also use the cli to verify including benchmarks.

This branch has conflicts that must be resolved
 to resolve conflicts before continuing.

src/retrieval/bm25.rs

File: src/retrieval/bm25.rs [hunk 1]
```diff
@@ -48,106 +48,88 @@ impl Default for Bm25Config {
...
-    /// Term frequencies in this document.
     term_freqs: HashMap<Arc<str>, u32>,
-    /// Document length (number of tokens).
     length: usize,
 }
 
-/// BM25 keyword search index.
-#[derive(Debug, Clone)]
+/// BM25-based document index for keyword search.
+#[derive(Debug, Default)]
```
[the remark this answers is not in the document]
Comment (d-o-hub) [reply src/retrieval/bm25.rs]: @jules address feedback and resolve the conversation

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -6,9 +6,6 @@ on:
   pull_request:
     branches: [ main ]
 
-permissions:
-  contents: read
-
 env:
```
[the remark this answers is not in the document]
Comment (d-o-hub) [reply .github/workflows/ci.yml:12]: @jules address feedback and resolve the conversation

**Comment by d-o-hub** [comment]
@codex 

> @jules compact / update learnings and check plans/ - agents.md and reference files if changes are needed. use dog food skill and verify in db and files save/load - also use the cli to verify including benchmarks.
> 
> This branch has conflicts that must be resolved to resolve conflicts before continuing.
> 
> src/retrieval/bm25.rs
